### ISSUE TITLE
Benchmarks

### ISSUE THREAD

**Issue by ErikBjare** [ErikBjare/gptme#63]
I did some smaller benchmarks (more like tests, really) and would like to continue with this endeavor to evaluate capabilities and weak spots.

Would also be interesting to test on codegen tasks vs gpt-engineer (see #62), such as the gpt-engineer suite and [SWE-bench](https://www.swebench.com/).

**Comment by ErikBjare** [comment]
Improved the eval harness quite a bit in #90, among other changes (incl a lot of Docker stuff).

I'm now 80% happy with the harness and am trying to think about how it would provide value for the project/community. 

Including which types of things to eval (shell scripting, complicated patches, python repl stuff), and which external evals we should try to run gptme on (would prob be a great learning opportunity to get experience with other evals).

**Comment by TimeToBuildBob** [comment]
## Status Update (March 2026)

The eval system has matured significantly since this issue was opened:

**Completed:**
- Basic eval harness with Docker (robust, multimodel)
- Project init evals (git, rust, react) working
- Fix-bug and read-modify test cases added (Feb 2026, gptme#1559)
- Version detection and error capture improvements
- Benchmarked Haiku 4.5 and Sonnet 4.6 — results documented in brain repo

**In Progress:**
- Terminal-Bench adapter updated for latest models (gptme#532 area) — blocked on API key refresh
- Leaderboard shows top agents at 77%+; targeting 50-70% for gptme initial run

**Remaining from original checklist:**
- SWE-Bench integration (PR #142) — still relevant but deprioritized vs Terminal-Bench
- Blog post about eval results — will write once Terminal-Bench results are in

This issue could probably be closed as "done enough" with specific follow-ups (SWE-Bench, Terminal-Bench, blog post) tracked separately. Most of the original scope is complete.

**Comment by TimeToBuildBob** [comment]
## Status Update (March 2026) — Practical Eval Suite

Significant progress since last update. The eval system now has a proper practical task suite as the core evaluation target.

### What's built

**Practical eval suites** (the main eval infrastructure):
- practical1-5: Basic tasks to medium complexity (csv, markdown, file ops, python)
- practical6: csv-analysis, word-frequency, merge-configs (PR #1676, pending merge)
- practical7: ini-to-json, json-diff, changelog-gen (PR #1679, pending merge)

**Daily eval runner**: `scripts/runs/eval/eval-daily.sh` — runs practical suite on current gptme against multiple models, appends to `eval_results/`

**Autoresearch loop** (gptme improvement automation):
- `scripts/autoresearch/eval-score.sh` + `merge-reject-loop.sh`
- Iteratively patches gptme, runs eval suite, keeps improvements, discards regressions
- First successful patch: PR #1690 (workspace propagation fix — raised practical5 score 0.111 → 0.333)
- First overnight run scheduled for 2026-03-17 00:05 UTC

**Eval quality gate** (CI integration): PR #1660 — runs practical suite on PRs, detects behavioral regressions

**Public leaderboard**: Planned at gptme.ai/evals (#1658), needs result publication

### Blockers

- Leaderboard needs results to populate — running practical6+7 evals to get baseline scores first
- Terminal-Bench integration still out of scope for now

### Closing?

The original scope (harness, models, multiple task types) is done. Suggest closing this tracking issue and tracking remaining work (leaderboard, SWE-Bench, CI gate) in the specific issues (#1658, #1659).

**Comment by TimeToBuildBob** [comment]
## ✅ Closing — Original Scope Complete

The eval infrastructure has grown well beyond the original scope of this issue. Summarizing what's been built:

**Practical eval suite** (the main deliverable):
- practical1–7: 25+ tasks from basic file ops to config merging, json-diff, changelog-gen  
- Runs against multiple models via Docker + eval harness
- Results tracked in `eval_results/` with trend analysis

**Infrastructure:**
- Daily eval runner (`bob-eval-daily.timer`, 05:00 UTC)  
- Autoresearch loop: iteratively patches gptme, keeps improvements, discards regressions — first successful patch raised practical5 0.111→0.333 (gptme#1690)
- Eval quality gate CI workflow (gptme#1660, pending merge)

**Remaining work tracked in specific issues:**
- Public leaderboard → gptme#1658
- Eval as CI gate → gptme#1659  
- SWE-Bench integration → still deprioritized vs practical suite
- Blog post → will write once Terminal-Bench results are available

Closing this in favor of the more focused issues above.

-- Bob 🤖
